text/0000-capture-disjoint-fields.md

```diff
@@ -0,0 +1,280 @@
+
+In a sense, when a closure is lowered to MIR, a list of "capture expressions" is
+created, which we will call the "capture set". Each expression is some part of
+the closure body which, in order to capture parts of the enclosing scope, must
+be pre-evaluated when the closure is created. The output of the expressions,
+which we will call "capture data", is stored in the anonymous struct which
+implements the `Fn*` traits. If a binding is used within a closure, at least one
+capture expression which borrows or moves that binding's value must exist in the
+capture set.
+
+Currently, lowering creates exactly one capture expression for each used
+binding, which borrows or moves the value in its entirety. This RFC proposes
+that lowering should instead create the minimal capture, where each expression
+is as specific as possible.
+
+This minimal set of capture expressions *might* be created by starting with the
+existing capture set (one maximal expression per binding) and then iterativly
+modifying and splitting the expressions by adding additional dereferences and
+path components.
+
+A capture expression is minimal if it produces...
+
+- a value that is not a struct or borrowed struct (e.g. primitive, enum, union).
+- a value that is used by the closure in its entirety (e.g. passed outside the closure).
+- any other case where the expression can not be made more specific (see below).
+
+Additionally capture expressions are not allowed to...
+
+- call impure functions.
+- make illegal moves (for example, out of a `Drop` type).
+- *(comments please, am I missing any cases?)*.
+
+Note that *all* functions are considered impure (including to overloaded deref
+impls). And, for the sake of capturing, all indexing is considered impure *(see
+unresolved)*. It is possible that overloaded `Deref::deref` implementations
+could be marked as pure by using a new, marker trait (such as `DerefPure`) or
+attribute (such as `#[deref_transparent]`). However, such a solution should be
+proposed in a separate RFC. In the meantime, `<Box as Deref>::deref` could be a
+special case of a pure function *(see unresolved)*.
+
+Note that, because capture expressions are all subsets of the closure body,
+this RFC does not change *what* is executed. It does change the order/number of
+executions for some operations, but since these must be pure, order/repetition
+does not matter. Only changes to lifetimes might be breaking. Specifically, the
+drop order of uncaptured data can be altered.
+
+We might solve this by considering a struct to be minimal if it contains unused
+fields that implement `Drop`. This would prevent the drop order of those fields
+from changing, but feels strange and non-orthogonal *(see unresolved)*.
+Encountering this case at all could trigger a warning, so that this extra rule
+could exist temporarily but be removed over the next epoc *(see unresolved)*.
+
+## Reference Examples
+
+Below are examples of various closures and their capture sets.
```

Comment:
(it might be good to revise the examples to add uses of the borrows, so that their meaning remains the same even when we move to non-lexical lifetimes. I know the intent is reasonably inferable, but still...)